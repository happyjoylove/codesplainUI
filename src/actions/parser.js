export const LOAD_PARSER = 'LOAD_PARSER';
export const ADD_ERROR = 'ADD_ERROR';
export const CLEAR_ERRORS = 'CLEAR_ERRORS';

export const loadParser = language => ({
  type: LOAD_PARSER,
  meta: {
    WebWorker: true,
  },
  payload: language,
});

export const addError = error => ({
  type: ADD_ERROR,
  payload: error,
});

export const clearErrors = () => ({
  type: CLEAR_ERRORS,
});
