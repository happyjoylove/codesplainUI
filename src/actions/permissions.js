export const SET_PERMISSIONS = 'SET_PERMISSIONS';
export const SET_AUTHOR = 'SET_AUTHOR';

export const setPermissions = permissions => ({
  type: SET_PERMISSIONS,
  payload: permissions,
});

export const setAuthor = author => ({
  type: SET_AUTHOR,
  payload: author,
});
