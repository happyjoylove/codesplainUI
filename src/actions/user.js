import axios from 'axios';
import mapValues from 'lodash/mapValues';

import { makeSaveEndpointUrl, sanitizeSnippetList } from '../util/requests';

const API_URL = process.env.REACT_APP_API_URL;

// Util func to check for 'NoSuchKey' responses from S3
const noSuchKey = data => (
  typeof data === 'string' && data.includes('NoSuchKey')
);

export const ADD_ORG = 'ADD_ORG';
export const ADD_ORGANIZATIONS = 'ADD_ORGANIZATIONS';
export const CLEAR_USER_CREDENTIALS = 'CLEAR_USER_CREDENTIALS';
export const RESTORE_USER_CREDENTIALS = 'RESTORE_USER_CREDENTIALS';
export const SAVE_ACCESS_TOKEN = 'SAVE_ACCESS_TOKEN';
export const SAVE_USERNAME = 'SAVE_USERNAME';
export const SET_AVATAR_URL = 'SET_AVATAR_URL';
export const SET_SNIPPET_LISTS = 'SET_SNIPPET_LISTS';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const SWITCH_ORG = 'SWITCH_ORG';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';

export const setAvatarUrl = url => ({
  type: SET_AVATAR_URL,
  payload: url,
});

export const addOrg = org => ({
  type: ADD_ORG,
  payload: org,
});

export const addOrganizations = organizations => ({
  type: ADD_ORGANIZATIONS,
  payload: organizations,
});

export const clearUserCredentials = () => ({
  type: CLEAR_USER_CREDENTIALS,
});

export const saveAccessToken = accessToken => ({
  type: SAVE_ACCESS_TOKEN,
  payload: accessToken,
});

export const saveUsername = username => ({
  type: SAVE_USERNAME,
  payload: username,
});

export const setSnippetLists = snippetLists => ({
  type: SET_SNIPPET_LISTS,
  payload: snippetLists,
});

export const setUserSnippets = snippetMeta => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const switchOrg = org => ({
  type: SWITCH_ORG,
  payload: org,
});

export const updateUserSnippets = () => (dispatch, getState) => {
  const {
    user: {
      token,
      username,
    },
  } = getState();
  // Fetch the user's snippet meta data and save it
  const headers = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios.get(makeSaveEndpointUrl(username), { headers })
    .then(({ data }) => {
      // Jump to catch block if the user has no index.json file:
      if (noSuchKey(data)) {
        throw new Error(`index.json does not exist for ${username}!`);
      }
      dispatch(setUserSnippets(data));
    });
};

export const fetchSnippetLists = () => (dispatch, getState) => {
  const {
    token,
    orgs,
  } = getState().user;
  const headers = {
    Authorization: `token ${token}`,
  };
  const url = `${API_URL}/users?users=${orgs.join(',')}`;
  // Remove the username from the org list
  return axios({
    method: 'GET',
    url,
    headers,
  })
    .then(({ data: snippetLists }) => {
      const sanitizedSnippetLists = mapValues(snippetLists, sanitizeSnippetList);
      dispatch(setSnippetLists(sanitizedSnippetLists));
    });
};

export const fetchAccessToken = authCode => () => {
  const reqUrl = `${API_URL}/auth`;
  const reqBody = { code: authCode };
  return axios({
    method: 'POST',
    url: reqUrl,
    data: reqBody,
  });
};

export const fetchUserInfo = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user',
    headers: reqHeaders,
  });
};

export const fetchUserOrgs = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user/orgs',
    headers: reqHeaders,
  });
};
