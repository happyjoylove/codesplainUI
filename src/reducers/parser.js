import * as actions from '../actions/parser';

const initialState = {
  language: '',
  errors: [],
};

const parser = (state = initialState, action) => {
  switch (action.type) {
  case actions.LOAD_PARSER: {
    if (action.meta) {
        // Missed the middleware somehow
      return state;
    }
    const language = action.payload;
    return { ...state, language };
  }
  case actions.ADD_ERROR: {
    const errors = state.errors.concat(action.payload);
    return {
      ...state,
      errors,
    };
  }
  case actions.CLEAR_ERRORS: {
    return {
      ...state,
      errors: [],
    };
  }
  default: {
    return state;
  }
  }
};

export default parser;
