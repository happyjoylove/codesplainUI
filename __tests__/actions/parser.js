import * as actions from '../../src/actions/parser';

describe('Actions: parser', () => {
  describe('LOAD_PARSER', () => {
    it('creates an action to load a parser', () => {
      const language = 'squanchy';
      expect(actions.loadParser(language)).toMatchSnapshot();
    });
  });
  describe('ADD_ERROR', () => {
    it('creates an action to add an error', () => {
      const error = {
        type: 'syntaxError',
        begin: 5,
        end: NaN,
        msg: 'blahblahblah',
      };
      expect(actions.addError(error)).toMatchSnapshot();
    });
    describe('CLEAR_ERRORS', () => {
      it('creates an action to remove all errors from state', () => {
        expect(actions.clearErrors()).toMatchSnapshot();
      });
    });
  });
});
