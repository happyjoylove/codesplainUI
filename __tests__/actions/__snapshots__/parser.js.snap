exports[`Actions: parser ADD_ERROR CLEAR_ERRORS creates an action to remove all errors from state 1`] = `
Object {
  "type": "CLEAR_ERRORS",
}
`;

exports[`Actions: parser ADD_ERROR creates an action to add an error 1`] = `
Object {
  "payload": Object {
    "begin": 5,
    "end": NaN,
    "msg": "blahblahblah",
    "type": "syntaxError",
  },
  "type": "ADD_ERROR",
}
`;

exports[`Actions: parser LOAD_PARSER creates an action to load a parser 1`] = `
Object {
  "meta": Object {
    "WebWorker": true,
  },
  "payload": "squanchy",
  "type": "LOAD_PARSER",
}
`;
